### DISCUSSION TITLE
Why would we separate illum model with analysis model

### DISCUSSION THREAD

**Discussion by zhangfan0319** [CellProfiler/CellProfiler#4633] [Q&A] [answered]
I am studying cellprofiler right now. Now, I'm wondering why all the pipelines includes the illum.cppipe pipeline and the analysis.cppipe pipeline. I checked the "illum.cppipe" file and found out the saveimage model only saved image in last cycle. So, how does the last image take information of all wells and why don't we gathered the illum pipeline with analysis pipeline

**Comment by bethac07** [comment, marked as answer]
Hi,

CellProfiler offers 3 ways to calculate the function - on each image (aka "each"), on all the loaded images but as part of the same pipeline that then goes on to run all the analyses (aka "first"), and on all the loaded images but as its own pipeline (aka "all"). The pipeline you're talking about uses "all" mode.

It's run separately for a couple of reasons 
- Why not "each": Because averaging across a whole plate's worth of data allows us to separate image-to-image illumination variance from the variance present in the light path, and that's what we care about. (Though sometimes we use "each" as an additional cleanup step in the downstream pipeline!)
- Why not "first", reason 1: Because if we ever need to re-analyze this plate with different settings, we'll have to repeat this calculation, and it takes a long time if you have hundreds or thousands of images! Better to run it once and then never again.
- Why not "first", reason 2: Because it means that we need to then run all our analyses in a single copy of CellProfiler that has a whole plate's worth of images loaded, which is probably fine if we have ie a 96 well plate with 2 sites per well, but is definitely NOT ideal for a 1536 well plate with 4 or more sites per well. We'll still always have to run the illumination correction itself on the whole plate, but doing it as a separate pipeline means segmentation and measurement, which are much slower than illumination correction , can be run on a cluster across lots of "nodes", which means the "clock-time" between our analysis starting and finishing will be a lot shorter and that one crash doesn't ruin a whole lot of analysis. 

Hope that helps!
